apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/asyncprofiler/AsyncProfilerDataSender.java

```diff
@@ -0,0 +1,190 @@
+        ServiceManager.INSTANCE.findService(GRPCChannelManager.class).addChannelListener(this);
+    }
+
+    @Override
+    public void boot() throws Throwable {
+
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+
+    }
+
+    @Override
+    public void statusChanged(GRPCChannelStatus status) {
+        if (GRPCChannelStatus.CONNECTED.equals(status)) {
+            Channel channel = ServiceManager.INSTANCE.findService(GRPCChannelManager.class).getChannel();
+            asyncProfilerTaskStub = AsyncProfilerTaskGrpc.newStub(channel);
+        } else {
+            asyncProfilerTaskStub = null;
+        }
+        this.status = status;
+    }
+
+    public void sendData(AsyncProfilerTask task, FileChannel channel) throws IOException, InterruptedException {
+        if (status != GRPCChannelStatus.CONNECTED || Objects.isNull(channel) || !channel.isOpen()) {
+            return;
+        }
+
+        int size = Math.toIntExact(channel.size());
+        boolean[] isAbort = new boolean[]{false};
+        final GRPCStreamServiceStatus status = new GRPCStreamServiceStatus(false);
+        StreamObserver<AsyncProfilerData> dataStreamObserver = asyncProfilerTaskStub.withDeadlineAfter(
+                GRPC_UPSTREAM_TIMEOUT, TimeUnit.SECONDS
+        ).collect(new StreamObserver<Commands>() {
+            @Override
+            public void onNext(Commands value) {
+
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                isAbort[0] = true;
+                status.finished();
+                if (LOGGER.isErrorEnable()) {
+                    LOGGER.error(
+                            t, "Send async profiler task data to collector fail with a grpc internal exception."
+                    );
+                }
+                ServiceManager.INSTANCE.findService(GRPCChannelManager.class).reportError(t);
+            }
+
+            @Override
+            public void onCompleted() {
+                status.finished();
+            }
+        });
+        AsyncProfilerMetaData metaData = AsyncProfilerMetaData.newBuilder()
+                .setService(Config.Agent.SERVICE_NAME)
+                .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                .setType(AsyncProfilerCollectType.PROFILING_SUCCESS)
+                .setContentSize(size)
+                .setTaskId(task.getTaskId())
+                .build();
+        AsyncProfilerData asyncProfilerData = AsyncProfilerData.newBuilder().setMetaData(metaData).build();
+        dataStreamObserver.onNext(asyncProfilerData);
+        /**
+         * Wait briefly to see if the server can receive the jfr. If the server cannot receive it, onError will be triggered.
+         * Then we wait for a while (waiting for the server to send onError) and then decide whether to send the jfr file.
+         */
+        Thread.sleep(500);
```

Comment:
@wu-sheng In the current design, the Java Agent has to wait after the first request in order to determine whether the OAP accepts the following data transfer. This is a kind of negotiation.

WDYT?

Reply:
`500ms` can't guarantee anything AFAIK.

Reply:
Note, in this case, Command should not be the response anymore.